src/rideables/SSHashTable.hpp

```diff
@@ -0,0 +1,322 @@
+#ifndef SSHASHTABLE_HPP
+#define SSHASHTABLE_HPP
+#include <stdio.h>
+#include <stdlib.h>
+
+#include <iostream>
+#include <atomic>
+#include <algorithm>
+#include <functional>
+#include <vector>
+#include <utility>
+
+#include "HarnessUtils.hpp"
+#include "ConcurrentPrimitives.hpp"
+#include "RMap.hpp"
+#include "RCUTracker.hpp"
+#include "CustomTypes.hpp"
+
+template <class K, class V>
+class SSHashTable : public RMap<K, V>
+{
+    struct Node;
+
+    struct MarkPtr
+    {
+        std::atomic<Node *> ptr;
+        MarkPtr(Node *n) : ptr(n){};
+        MarkPtr() : ptr(nullptr){};
+    };
+
+    struct Node
+    {
+        K key;
+        V val;
+        MarkPtr next;
+        Node(K k, V v, Node *n) : key(k), val(v), next(n){};
+    };
+
+private:
+    int size = 1000000; //number of buckets for hash table
+    int MAX_LOAD = 100000000;
+    padded<MarkPtr> *buckets = new padded<MarkPtr>[size] {};
+    atomic<int32_t> count = {0};
+    RCUTracker<Node> tracker;
+    //TODO : must be thread private
+    MarkPtr *prev = nullptr;
+    Node *curr = nullptr;
+    Node *next = nullptr;
+
+    const uint64_t MARK_MASK = ~0x1;
+
+    inline Node *getPtr(Node *mptr)
+    {
+        return (Node *)((uint64_t)mptr & MARK_MASK);
+    }
+    inline bool getMark(Node *mptr)
+    {
+        return (bool)((uint64_t)mptr & 1);
+    }
+    inline Node *mixPtrMark(Node *ptr, bool mk)
+    {
+        return (Node *)((uint64_t)ptr | mk);
+    }
+    inline Node *setMark(Node *mptr)
+    {
+        return mixPtrMark(mptr, true);
+    }
+
+public:
+    SSHashTable(int task_num) : tracker(task_num, 100, 1000, true){};
+    ~SSHashTable(){};
+
+    void initialize_bucket(int bucket, int tid);
+    int get_parent(int bucket);
+    optional<K> reverse_bits(K n);
+    optional<K> so_regularkey(K key);
+    optional<K> so_dummykey(K key);
+    optional<V> get(K key, int tid);
+    bool remove(K key, int tid);
+    bool put(K key, V val, int tid);
```

Comment:
`put` and `remove` return `optional<T>`. Also, `RMap<T>` has another two pure virtual functions `bool insert(K key, V val, int tid)` and `optional<V> replace(K key, V val, int tid)` that have to be implemented in its derived class, or it won't compile. Check out the definition of the base class `RMap<T>` in [RMap.hpp](https://github.com/urcs-sync/Montage/blob/master/src/RMap.hpp).